src/Datastore/Operation.php

```diff
@@ -0,0 +1,659 @@
+     * }
+     * @return array
+     * @codingStandardsIgnoreEnd
+     */
+    public function lookup(array $keys, array $options = [])
+    {
+        $options += [
+            'className' => null
+        ];
+
+        $this->validateBatch($keys, Key::class, function ($key) {
+            if ($key->state() !== Key::STATE_COMPLETE) {
+                throw new InvalidArgumentException(sprintf(
+                    'Given $key is in an invalid state. Can only lookup records when given a complete key. ' .
+                    'Given path was %s',
+                    (string) $key
+                ));
+            }
+        });
+
+        $res = $this->connection->lookup($options + [
+            'projectId' => $this->projectId,
+            'readOptions' => $this->readOptions($options),
+            'keys' => $keys
+        ]);
+
+        $result = [];
+        if (isset($res['found'])) {
+            $result['found'] = $this->mapEntityResult(
+                $res['found'],
+                $options['className']
+            );
+        }
+
+        if (isset($res['missing'])) {
+            $result['missing'] = [];
+            foreach ($res['missing'] as $missing) {
+                $key = $this->key(
+                    $missing['entity']['key']['path'],
+                    $missing['entity']['key']['partitionId']
+                );
+
+                $result['missing'][] = $key;
+            }
+        }
+
+        if (isset($res['deferred'])) {
+            $result['deferred'] = [];
+            foreach ($res['deferred'] as $deferred) {
+                $key = $this->key(
+                    $deferred['path'],
+                    $deferred['partitionId']
+                );
+
+                $result['deferred'][] = $key;
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Run a query and return entities
+     *
+     * @param QueryInterface $query The query object.
+     * @param array $options {
+     *     Configuration Options
+     *
+     *     @type string $transaction The transaction ID, if the query should be
+     *           run in a transaction.
+     *     @type string $className The name of the class to return results as.
+     *           Must be a subclass of {@see Google\Cloud\Datastore\Entity}.
+     *           If not set, {@see Google\Cloud\Datastore\Entity} will be used.
+     * }
+     * @return \Generator<Google\Cloud\Datastore\Entity>
+     */
+    public function runQuery(QueryInterface $query, array $options = [])
+    {
+        $options += [
+            'transaction' => null,
+            'className' => null
+        ];
+
+        $moreResults = true;
+        do {
+            $res = $this->connection->runQuery($options + [
+                'projectId' => $this->projectId,
+                'partitionId' => $this->partitionId($this->projectId, $this->namespaceId),
+                'readOptions' => $this->readOptions($options),
```

Comment:
When we start validating against the schemas there is a good chance we will end up throwing exceptions for keys that shouldn't exist in a request, so yes if we can catch and remove things like this now that would be great.